### PR TITLE
merge: Add support for merging update diffs into an existing object.

### PR DESCRIPTION
**Pull request by kav-ya**
Test Plan:
Minimal tests added; more upcoming.

### CODE REVIEW

File: merge/merge.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	case map[string]interface{}:
+		new = MergeMap(prev, d)
+	case []interface{}:
+		new = MergeArray(prev, d)
+	}
+	return new
+}
+
+// MergeMap applies a delta to a map.
+func MergeMap(prev map[string]interface{}, diff map[string]interface{}) map[string]interface{} {
```
Comment (jellevandenhooff) [inline merge/merge.go]: These can be private.

File: merge/merge.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	for k, d := range diff {
+		if _, ok := prev[k]; !ok {
+			new[k] = MergeReplaced(d)
+		}
+	}
+
+	return new
+}
+
+// MergeMap applies a delta to an array.
```
Comment (jellevandenhooff) [inline merge/merge.go]: MergeArray.

File: merge/merge.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
+package diff
```
Comment (jellevandenhooff) [inline merge/merge.go]: merge

File: merge/merge_test.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
+package diff
```
Comment (jellevandenhooff) [inline merge/merge_test.go]: merge_test

File: merge/merge_test.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+import (
+	"reflect"
+	"testing"
+
+	"thunder/internal"
```
Comment (jellevandenhooff) [inline merge/merge_test.go]: github.com/samsarahq/thunder/internal

File: merge/merge_test.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+		},
+		{
+			Case:        "Complex field",
+			Prev:        `{"friends": ["alice", "charlie"]}`,
+			Diff:        `{"friends": [["eli"]]}`,
+			ExpectedNew: `{"friends": ["eli"]}`,
+		},
+	}
+
+	runTests(t, cases)
```
Comment (jellevandenhooff) [inline merge/merge_test.go]: You can now merge all these functions and inline runTests.

File: merge/merge.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+func Merge(prev interface{}, diff interface{}) interface{} {
+	d, ok := diff.(map[string]interface{})
+	if !ok {
+		return MergeReplaced(diff)
+	}
+
+	var new interface{}
+	switch prev := prev.(type) {
+	case map[string]interface{}:
+		new = MergeMap(prev, d)
```
Comment (jellevandenhooff) [inline merge/merge.go]: How about returning directly here, adding a default return nil?

File: merge/merge_test.go (hunk 5 above)
Comment (kav-ya) [inline merge/merge_test.go]: Whoops, thanks.

File: merge/merge.go (hunk 1 above)
Comment (kav-ya) [inline merge/merge.go]: Whoops, thought I did this; updated.

File: merge/merge.go (hunk 3 above)
Comment (kav-ya) [inline merge/merge.go]: done.

File: merge/merge.go (hunk 2 above)
Comment (kav-ya) [inline merge/merge.go]: done.

File: merge/merge_test.go (hunk 4 above)
Comment (kav-ya) [inline merge/merge_test.go]: done.

File: merge/merge_test.go (hunk 6 above)
Comment (kav-ya) [inline merge/merge_test.go]: done.

File: merge/merge.go (hunk 7 above)
Comment (kav-ya) [inline merge/merge.go]: done.

**Comment by kav-ya** [comment]
Updated @jellevandenhooff, PTAL.

File: merge/merge_test.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+		{
+			Case:        "Map",
+			Prev:        `{"name": "bob", "address": {"state": "ca", "city": "sf"}, "age": 30}`,
+			Diff:        `{"name": "alice", "address": {"city": "oakland"}, "age": [], "friends": [["bob", "charlie"]]}`,
+			ExpectedNew: `{"name": "alice", "address": {"state": "ca", "city": "oakland"}, "friends": ["bob", "charlie"]}`,
+		},
+	}
+
+	for _, c := range cases {
+		new := Merge(internal.ParseJSON(c.Prev), internal.ParseJSON(c.Diff))
```
Comment (jellevandenhooff) [inline merge/merge_test.go]: I think this should be merge.Merge (see CI failure).

Comment (kav-ya) [inline merge/merge_test.go]: :/; fixed.

**Pull request merged**